src/middlewared/middlewared/api/v26_04_0/container.py

```diff
@@ -0,0 +1,164 @@
+from typing import Literal
+
+from pydantic import Field, PositiveInt
+
+from middlewared.api.base import (
+    BaseModel, Excluded, excluded_field, ForUpdateMetaclass, NonEmptyString, single_argument_args, UUIDv4String,
+)
+
+
+__all__ = [
+    "ContainerEntry",
+    "ContainerCreateArgs", "ContainerCreateResult",
+    "ContainerUpdateArgs", "ContainerUpdateResult",
+    "ContainerDeleteArgs", "ContainerDeleteResult",
+    "ContainerStartArgs", "ContainerStartResult",
+    "ContainerStopArgs", "ContainerStopResult",
+    "ContainerMigrateArgs", "ContainerMigrateResult",
+]
+
+
+class IdmapConfigurationItem(BaseModel):
+    target: PositiveInt
+    """UID/GID 0 in container will be mapped to this target UID/GID in host."""
+    count: PositiveInt
+    """How many users/groups in container are allowed to map to host"s user/group."""
+
+
+class IdmapConfiguration(BaseModel):
+    uid: IdmapConfigurationItem
+    """UID mapping configuration."""
+    gid: IdmapConfigurationItem
+    """GID mapping configuration."""
+
+
+class ContainerStatus(BaseModel):
+    state: Literal["RUNNING", "STOPPED"]
+    """Container state."""
+    pid: int | None
+    """Container PID (if running)."""
+    domain_state: NonEmptyString | None
+    """Domain state reported by libvirt."""
+
+
+class ContainerEntry(BaseModel):
+    id: int
+    """Container ID."""
+    uuid: UUIDv4String | None = None
+    """Container UUID (for libvirt)."""
+    name: NonEmptyString
+    """Container name."""
+    description: str = ""
+    """Container description."""
+    vcpus: int | None = Field(ge=1, default=None)
+    """How many CPUs container can use."""
+    cores: int | None = Field(ge=1, default=None)
+    """How many cores does each CPU have."""
+    threads: int | None = Field(ge=1, default=None)
+    """How many threads does each CPU core have."""
+    cpuset: str | None = None  # TODO: Add validation for numeric set
+    """List of physical CPU numbers that domain process and virtual CPUs can be pinned to by default."""
+    memory: int | None = Field(ge=20, default=None)
+    """Memory available to container (in megabytes)."""
+    autostart: bool = True
+    """Automatically start the container on boot."""
+    time: Literal["LOCAL", "UTC"] = "LOCAL"
+    """Whether container time should be local time or UTC time."""
+    shutdown_timeout: int = Field(ge=5, le=300, default=90)
+    """How many seconds to wait for container to shut down before killing it."""
+    dataset: str
+    """Which dataset to use as the container root filesystem."""
```

Comment:
Should we do any validation here on dataset names? Are we restricting the root filesystem of container to anywhere special (like child of `tank/ix-virt`)?

Reply:
No, why would we want that?